lte/cloud/go/sms_ll/sms_rp.go

```diff
@@ -0,0 +1,367 @@
+	b := make([]byte, 0, 11)
+	b = append(b, byte(rpadde_len))
+	if rpadde.Length != 0x0 {
+		rpadde_len += 1 + len(rpadde.Number)
+		b = append(b, rpadde.NumberInfo)
+		b = append(b, rpadde.Number...)
+		b[0] = byte(rpadde_len) // update with new size
+		if rpadde_len != int(rpadde.Length) {
+			return nil, SMSRPError("Address size mismatch")
+		}
+	}
+	return b, nil
+}
+
+// Decode an address element. Returns the length of the address element if present.
+func (rpadde *RPAddressElement) UnmarshalBinary(input []byte) (int, error) {
+	// Empty addresses will be one byte long with a zero value length
+	if len(input) == 1 {
+		if input[0] != 0x0 {
+			return -1, SMSRPError("Invalid RP Address of length 1")
+		} else {
+			rpadde.Length = input[0]
+			return 1, nil
+		}
+	} else if len(input) < 2 {
+		return -1, SMSRPError("Invalid RP Address")
+	}
+
+	rpadde.Length = input[0]
+
+	rpadde.NumberInfo = input[1]
+	rpadde.Number = input[2:int(rpadde.Length)]
+	return int(rpadde.Length), nil
+}
+
+// The RP-Address-Element refers to the number of the SMSC. In our case, we
+// generally don't use an SMSC, so we just set the number to 11.
+func (rpadde *RPAddressElement) SetFakeNumber() error {
+	rpadde.NumberInfo = 0xb9     // network specific number, private numbering plan
+	rpadde.Number = []byte{0x11} // the decimal number 1 1
+	rpadde.Length = 0x2
+
+	return nil
+}
+
+// RP-User data element (TS 24.011 8.2.5.3)
+type RPUserElement struct {
+	IEI    byte // Not present for RP-DATA
+	Length byte
+	TPDU   []byte
+}
+
+func CreateRPUserElement(data []byte) (*RPUserElement, error) {
+	rpue := new(RPUserElement)
+	rpue.IEI = RP_UDE_IEI
+	if len(data) > 232 { // TS24.011 8.2.5.3
+		return nil, SMSRPError("UserData-Element too long (>232 bytes)")
+
+	}
+	rpue.Length = byte(len(data))
+	rpue.TPDU = data
+
+	return rpue, nil
+}
+
+func (rpue RPUserElement) MarshalBinary(msgType byte) ([]byte, error) {
+	rpu_len := len(rpue.TPDU) + 1
+	b := make([]byte, 0, rpu_len)
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		b = append(b, rpue.IEI)
+	}
+	b = append(b, rpue.Length)
+	b = append(b, rpue.TPDU...)
+	return b, nil
+}
+
+func (rpue *RPUserElement) UnmarshalBinary(msgType byte, input []byte) (int, error) {
+	idx := 0
+	if msgType == RP_ACK || msgType == RP_ERROR { // these start with IEI
+		rpue.IEI = input[idx]
+		idx++
+	}
+	rpue.Length = input[idx]
+	idx++
+	n := idx + int(rpue.Length)
+	rpue.TPDU = input[idx:n]
+	idx += n
+	return n, nil
+}
+
+// RP-Cause element (TS 24.011 8.2.5.4)
+type RPCauseElement struct {
+	IEI        byte // Never serialized
+	Length     byte
+	Cause      byte
+	Diagnostic byte // Optional
+}
+
+func (rpce RPCauseElement) MarshalBinary() ([]byte, error) {
+	b := make([]byte, 0, 2)
+	b = append(b, rpce.Length)
+	b = append(b, rpce.Cause)
+	if int(rpce.Length) == 2 {
+		b = append(b, rpce.Diagnostic)
+	}
+	return b, nil
+}
+
+func (rpce *RPCauseElement) UnmarshalBinary(input []byte) (int, error) {
```

Comment:
Do you specifically want to implement `BinaryMarshaler`/`BinaryUnmarshaler`? If not I would ditch the error return in cases like this when it's never non-nil